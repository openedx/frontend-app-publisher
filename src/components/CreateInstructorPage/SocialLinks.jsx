import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';

import RenderInputTextField from '../RenderInputTextField';
import RenderSelectField from '../RenderSelectField';
import RemoveButton from '../../components/RemoveButton';


class SocialLinks extends React.Component {
  constructor(props) {
    super(props);

    this.handleRemove = this.handleRemove.bind(this);
  }

  getSocialOptions() {
    return [
      { label: 'Please select a type', value: '' },
      { label: 'Facebook', value: 'facebook' },
      { label: 'Twitter', value: 'twitter' },
      { label: 'Blog', value: 'blog' },
      { label: 'Other', value: 'others' },
    ];
  }

  handleRemove(index) {
    this.props.fields.remove(index);
  }

  render() {
    const { fields } = this.props;

    return (
      <div className="social-links mt-4">
        <button
          type="button"
          className="btn btn-light js-add-button"
          onClick={() => fields.push({})}
        >
          Add social link
        </button>
        <ul className="list-group p-0 m-0">
          {fields.map((link, index) => (
            <li className="social-link list-group-item row d-flex align-items-center px-0 mx-0" key={link}>
              <div className="col-6">
                <Field
                  name={`${link}.type`}
                  component={RenderSelectField}
                  options={this.getSocialOptions()}
                  label={
                    <React.Fragment>
                      Type
                      <span className="required">*</span>
                    </React.Fragment>
                  }
                  required
                />
              </div>
              <div className="col-6">
                <Field
                  name={`${link}.title`}
                  component={RenderInputTextField}
                  type="text"
                  label={
                    <React.Fragment>
                      Title - optional
                    </React.Fragment>
                  }
                />
              </div>
              <div className="col-4">
                <Field
                  name={`${link}.url`}
                  component={RenderInputTextField}
                  type="text"
                  label={
                    <React.Fragment>
                      URL
                      <span className="required">*</span>
                    </React.Fragment>
                  }
                  required
                />
              </div>
              <RemoveButton
                label="Remove social link"
                onRemove={this.handleRemove}
                targetFieldNumber={index}
              />
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

SocialLinks.propTypes = {
  fields: PropTypes.shape({
    remove: PropTypes.func,
  }).isRequired,
};

export default SocialLinks;
