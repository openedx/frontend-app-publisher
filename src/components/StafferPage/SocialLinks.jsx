import React from 'react';
import PropTypes from 'prop-types';
import { Field, FieldArray } from 'redux-form';

import RenderInputTextField from '../RenderInputTextField';
import RenderSelectField from '../RenderSelectField';
import RemoveButton from '../RemoveButton';
import FieldLabel from '../FieldLabel';

class SocialLinks extends React.Component {
  constructor(props) {
    super(props);

    this.handleRemove = this.handleRemove.bind(this);
  }

  getSocialOptions() {
    return [
      { label: 'Please select a type', value: '' },
      { label: 'Facebook', value: 'facebook' },
      { label: 'Twitter', value: 'twitter' },
      { label: 'Blog', value: 'blog' },
      { label: 'Other', value: 'others' },
    ];
  }

  handleRemove(index) {
    this.props.fields.remove(index);
  }

  render() {
    const { fields } = this.props;

    return (
      <div className="social-links mb-3">
        <ul className="list-group p-0 m-0 container-fluid">
          {fields.map((link, index) => (
            <li className="social-link list-group-item row d-flex align-items-center px-0 mx-0" key={link}>
              <div className="col-4">
                <Field
                  name={`${link}.type`}
                  component={RenderSelectField}
                  options={this.getSocialOptions()}
                  label={<FieldLabel text="Type" required />}
                  required
                />
              </div>
              <div className="col-3">
                <Field
                  name={`${link}.title`}
                  component={RenderInputTextField}
                  type="text"
                  label={<FieldLabel text="Link Text" />}
                />
              </div>
              <div className="col-4">
                <Field
                  name={`${link}.url`}
                  component={RenderInputTextField}
                  type="url"
                  label={<FieldLabel text="URL" required />}
                  required
                />
              </div>
              <input
                name={`${link}.id`}
                type="hidden"
              />
              <RemoveButton
                className="col-1 m-auto"
                label="Remove social link"
                onRemove={this.handleRemove}
                targetFieldNumber={index}
              />
            </li>
          ))}
        </ul>
        <button
          type="button"
          className="btn btn-outline-primary js-add-button mt-2"
          onClick={() => fields.push({})}
        >
          Add social link
        </button>
      </div>
    );
  }
}

SocialLinks.propTypes = {
  fields: PropTypes.instanceOf(FieldArray).isRequired,
};

export default SocialLinks;
