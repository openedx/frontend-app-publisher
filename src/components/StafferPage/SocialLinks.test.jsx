import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import SocialLinks from './SocialLinks';

/*
*  Disable console errors for this test file so that we don't receive warnings
*  about fields being an array rather than an object. This prop change is
*  intentional as redux-form field arrays treat fields strangely:
*  'The fields object is a "pseudo-array", in that it has many of the same
*  properties and methods as a javascript Array, providing both reading and
*  writing functionality.'
*/
jest.spyOn(global.console, 'error').mockImplementation(() => jest.fn());

describe('Social links', () => {
  it('renders correctly with no fields', () => {
    const component = shallow(<SocialLinks fields={[]} />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('renders correctly when given fields', () => {
    const component = shallow(<SocialLinks fields={[{}]} />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('adds fields when the add button is pushed', () => {
    const fields = [{}];
    const component = shallow(<SocialLinks fields={fields} />);
    expect(fields.length).toEqual(1);

    component.find('.js-add-button').simulate('click');
    expect(fields.length).toEqual(2);
  });
});
