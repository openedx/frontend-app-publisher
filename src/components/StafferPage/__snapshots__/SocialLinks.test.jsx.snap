// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Social links renders correctly when given fields 1`] = `
<div
  className="social-links mb-3"
>
  <ul
    className="list-group p-0 m-0 container-fluid"
  >
    <li
      className="social-link list-group-item row d-flex align-items-center px-0 mx-0"
      key="[object Object]"
    >
      <div
        className="col-4"
      >
        <Field
          component={[Function]}
          label={
            <FieldLabel
              className=""
              extraText=""
              helpText=""
              id={null}
              optional={false}
              required={true}
              text="Type"
            />
          }
          name="[object Object].type"
          options={
            Array [
              Object {
                "label": "Please select a type",
                "value": "",
              },
              Object {
                "label": "Facebook",
                "value": "facebook",
              },
              Object {
                "label": "Twitter",
                "value": "twitter",
              },
              Object {
                "label": "Blog",
                "value": "blog",
              },
              Object {
                "label": "Other",
                "value": "others",
              },
            ]
          }
          required={true}
        />
      </div>
      <div
        className="col-3"
      >
        <Field
          component={[Function]}
          label={
            <FieldLabel
              className=""
              extraText=""
              helpText=""
              id={null}
              optional={false}
              text="Link Text"
            />
          }
          name="[object Object].title"
          type="text"
        />
      </div>
      <div
        className="col-4"
      >
        <Field
          component={[Function]}
          label={
            <FieldLabel
              className=""
              extraText=""
              helpText=""
              id={null}
              optional={false}
              required={true}
              text="URL"
            />
          }
          name="[object Object].url"
          required={true}
          type="url"
        />
      </div>
      <input
        name="[object Object].id"
        type="hidden"
      />
      <RemoveButton
        className="col-1 m-auto"
        label="Remove social link"
        onRemove={[Function]}
        targetFieldNumber={0}
      />
    </li>
  </ul>
  <button
    className="btn btn-outline-primary js-add-button mt-2"
    onClick={[Function]}
    type="button"
  >
    Add social link
  </button>
</div>
`;

exports[`Social links renders correctly with no fields 1`] = `
<div
  className="social-links mb-3"
>
  <ul
    className="list-group p-0 m-0 container-fluid"
  />
  <button
    className="btn btn-outline-primary js-add-button mt-2"
    onClick={[Function]}
    type="button"
  >
    Add social link
  </button>
</div>
`;
